### PR TITLE
FEAT - Split Lab build into an action of its own

### LINKED ISSUE #27: ENH -  Pull JupyterLab build and install steps into reusable action
Currently there are a few steps in .github/actions/jlab-a11y-test/action.yml but they could be pulled out into a separate action.

### PR DESCRIPTION
**Pull request by trallard**
Closes #27 

## TODO as part of this PR

- [x] It should be updated once #29  is merged into `main`
- [x] Add option to use an `SHA` for Lumino build/link in case this is needed
- [x] Update comments and docs to match the changes

🎉 I now modified the PR so that we can take the repo and `ref` of almost any package to build and test with JupyterLab

This run combines `jupyterlab/jupyterlab` with the branch from mark's lumino's PR that @gabalafou  is currently reviewing (`scmmmh/lumino`) https://github.com/Quansight-Labs/jupyter-a11y-testing/actions/runs/3961294989

### CODE REVIEW

**Comment by trallard** [comment]
That summary formatting is wonky, so I might need to figure out a way to deal with it... not urgent tho

File: .github/actions/link-package/action.yml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        repository: ${{ inputs.external_pkg_repo }}
+        ref: ${{ inputs.external_pkg_ref }}
+        path: external-pkg
+
+    # following https://github.com/jupyterlab/lumino/blob/main/CONTRIBUTING.md#setting-up-a-development-environment
+    - name: Build package from source 🏗
+      run: |
+        yarn
+        yarn build:src
+        conda run --prefix $CONDA_PREFIX jlpm link
```
Comment (gabalafou) [inline .github/actions/link-package/action.yml]: Do we need to add some note somewhere in this file that this action needs to be run in a certain context, like within a workflow that sets up `yarn` and `conda`? I mean, like, some way to indicate that this is very much not a standalone action.

File: .github/actions/link-package/action.yml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      shell: bash -l {0}
+      working-directory: external-pkg
+
+    - name: Link JupyterLab and package 🔗
+      run: |
+        # todo later on: replace the lumino reference here
+        conda run --prefix $CONDA_PREFIX jlpm link "lumino-top-level"
```
Comment (gabalafou) [inline .github/actions/link-package/action.yml:51]: Does this work? I'm not sure but I think you actually have to go through each of the (sub)packages in the Lumino monorepo and link them one by one. 

Is that what your "todo" comment is about?

File: .github/workflows/accessibility-test-jupyterlab.yml [hunk 3]
```diff
@@ -7,35 +7,73 @@ on:
...
+        description: |
+          External package repository to link, for example jupyter/lumino
+          only to be used if you want to link and test a package against JupyterLab.
+          Otherwise leave blank.
+        type: string
+        required: false
+        default: ""
+      external_pkg_ref:
+        description: |
+          Only to be used if you need to link a specific Lumino's ref (branch, SHA),
```
Comment (gabalafou) [inline .github/workflows/accessibility-test-jupyterlab.yml]: This description is not the same as the other 😛

**Review by gabalafou — APPROVED** [review]

**Comment by gabalafou** [comment]
Approved for now. If anything's not working we can fix it in subsequent pull requests.

**Comment by trallard** [comment]
I am going to merge this and if needed work on smaller self-contained improvements

File: .github/actions/link-package/action.yml (hunk 2 above)
Comment (trallard) [reply .github/actions/link-package/action.yml:51]: This works as is - will make improvements for robustness in a later iteration

Comment (gabalafou) [reply .github/actions/link-package/action.yml:51]: Yeah, I remember we chatted about this with @steff456, I think, and I may have just been flat out wrong about needing to link every sub-package versus the top-level package.

File: .github/actions/link-package/action.yml (hunk 1 above)
Comment (trallard) [reply .github/actions/link-package/action.yml]: Added now

**Comment by trallard** [comment]
Note I had to upgrade to node >= 18 on our deps (ljlab) since now whatever is in `master` dropped support for node16

File: .github/actions/link-package/action.yml [hunk 4]
```diff
@@ -30,17 +34,20 @@ runs:
...
+    # and https://jupyterlab.readthedocs.io/en/latest/developer/contributing.html#id17
     - name: Build package from source 🏗
       run: |
-        yarn
-        yarn build:src
-        conda run --prefix $CONDA_PREFIX jlpm link
+        echo "::group::Build package from source 🛠"
+        conda run --prefix $CONDA_PREFIX yarn
+        conda run --prefix $CONDA_PREFIX yarn build:src
+        conda run --prefix $CONDA_PREFIX YARN_IGNORE_PATH=1 yarn link
```
Comment (gabalafou) [inline .github/actions/link-package/action.yml:43]: `YARN_IGNORE_PATH` 😵‍💫? Why not use `jlpm link` here?

File: .github/actions/link-package/action.yml [hunk 5]
```diff
@@ -30,17 +34,20 @@ runs:
...
+        echo "::endgrup::"
```
Comment (gabalafou) [inline .github/actions/link-package/action.yml]:
```suggestion
        echo "::endgroup::"
```

File: .github/workflows/accessibility-test-jupyterlab.yml [hunk 6]
```diff
@@ -104,10 +105,10 @@ jobs:
     steps:
       # We have to check out the repo in order to locally reference the action,
       # i.e., .github/actions/...
-      - name: Check out repository 🛎
+      - name: Check out JLab testing repository 🛎
```
Comment (gabalafou) [inline .github/workflows/accessibility-test-jupyterlab.yml]: This feels like a minor thing, but I think "JLab testing repository" risks creating confusion than just "repository" (which I think conventionally in GitHub Actions means "**this** repository")

File: README.md [hunk 7]
```diff
@@ -40,6 +39,30 @@ We are starting with the web app UI, then we will add tests for the docs.
...
+This repository is organized as follows:
+
+```txt
+.
+├── .github
+│   ├── .github/workflows                               # set of GitHub actions to run the accessibility tests based on certain type of triggers
+│   └── .github/actions                                 # composable actions that perform specific tasks (not to be used on their own but as part of a GitHub actions workflow)
+├── testing                                             # root folder for the testing tools
+│   └── testing/jupyterlab                              # testing tools and scripts for JupyterLab
+│       ├── testing/jupyterlab/manual-testing-scripts   # "recipes" to perform manual accessibility checks on JupterLab (to be used in conjunction with the automated tests)
```
Comment (gabalafou) [inline README.md]:
```suggestion
│       ├── testing/jupyterlab/manual-testing-scripts   # "recipes" that explain in plain language how automated tests can be also be carried out manually
```

Comment (gabalafou) [reply README.md]: This makes me think there's something wrong with the naming or information hierarchy. This is why I originally named the directory something like test-info-pages... In my mental model, every accessibility test in this repo should come with its own human-readable page that explains what the test is, why it matters, how to carry out the test manually, and maybe other info.

File: README.md [hunk 8]
```diff
@@ -40,6 +39,30 @@ We are starting with the web app UI, then we will add tests for the docs.
...
+│       ├── testing/jupyterlab/tests                    # set of Playwright automated tests
+│       ├── testing/jupyterlab/environment.yml          # conda environment file to install the dependencies for the automated tests
+│       ├── testing/jupyterlab/README.md                # documentation for the JupyterLab tests - start here to learn how to run the tests locally or in GitHub actions
+│       ├── testing/jupyterlab/package.json             # npm package file to install the dependencies for the automated tests
+│       └── testing/jupyterlab/playwright.config.ts     # Playwright configuration file
+│   └── testing/notebooks                               # set of reference Jupter notebooks to be used in the automated tests
+│   └── testing/scripts                               # set of manual testing scripts for JupyterLab (include relevant WCAG success criteria and step-by-step guides to audit)
+├── .pre-comMit-config.yaml                             # configuration file for the pre-commit hooks
```
Comment (gabalafou) [inline README.md]:
```suggestion
├── .pre-commit-config.yaml                             # configuration file for the pre-commit hooks
```

File: README.md [hunk 9]
```diff
@@ -40,6 +39,30 @@ We are starting with the web app UI, then we will add tests for the docs.
...
+│   └── testing/jupyterlab                              # testing tools and scripts for JupyterLab
+│       ├── testing/jupyterlab/manual-testing-scripts   # "recipes" to perform manual accessibility checks on JupterLab (to be used in conjunction with the automated tests)
+│       ├── testing/jupyterlab/tests                    # set of Playwright automated tests
+│       ├── testing/jupyterlab/environment.yml          # conda environment file to install the dependencies for the automated tests
+│       ├── testing/jupyterlab/README.md                # documentation for the JupyterLab tests - start here to learn how to run the tests locally or in GitHub actions
+│       ├── testing/jupyterlab/package.json             # npm package file to install the dependencies for the automated tests
+│       └── testing/jupyterlab/playwright.config.ts     # Playwright configuration file
+│   └── testing/notebooks                               # set of reference Jupter notebooks to be used in the automated tests
+│   └── testing/scripts                               # set of manual testing scripts for JupyterLab (include relevant WCAG success criteria and step-by-step guides to audit)
```
Comment (gabalafou) [inline README.md]:
```suggestion
│   └── testing/scripts                                 # set of manual testing scripts for JupyterLab (include relevant WCAG success criteria and step-by-step guides to audit)
```

File: README.md [hunk 10]
```diff
@@ -40,6 +39,30 @@ We are starting with the web app UI, then we will add tests for the docs.
...
+├── testing                                             # root folder for the testing tools
+│   └── testing/jupyterlab                              # testing tools and scripts for JupyterLab
+│       ├── testing/jupyterlab/manual-testing-scripts   # "recipes" to perform manual accessibility checks on JupterLab (to be used in conjunction with the automated tests)
+│       ├── testing/jupyterlab/tests                    # set of Playwright automated tests
+│       ├── testing/jupyterlab/environment.yml          # conda environment file to install the dependencies for the automated tests
+│       ├── testing/jupyterlab/README.md                # documentation for the JupyterLab tests - start here to learn how to run the tests locally or in GitHub actions
+│       ├── testing/jupyterlab/package.json             # npm package file to install the dependencies for the automated tests
+│       └── testing/jupyterlab/playwright.config.ts     # Playwright configuration file
+│   └── testing/notebooks                               # set of reference Jupter notebooks to be used in the automated tests
```
Comment (gabalafou) [inline README.md]:
```suggestion
│   └── testing/notebooks                               # set of reference Jupyter notebooks to be used in the automated tests
```

File: README.md (hunk 9 above)
Comment (gabalafou) [reply README.md]: We should probably delete `testing/scripts` lest it cause confusion with `testing/jupyterlab/manual-testing-scripts`.

Or we need to decide on the path forward for this repo and our work.

**Review by gabalafou — CHANGES_REQUESTED** [review]
I **love** that we are going to have a way to build a specific version of Lumino into JupyterLab with these changes.

I caught a couple small errors that should probably be fixed before merging.

I also have some comments that shouldn't block this.

File: .github/actions/link-package/action.yml (hunk 4 above)
Comment (trallard) [reply .github/actions/link-package/action.yml:43]: I was using `jlpm link` but started failing with

```
ERROR conda.cli.main_run:execute(47): `conda run jlpm link` failed. (See above for error)
Unknown Syntax Error: Not enough positional arguments.

$ yarn link [-A,--all] [-p,--private] [-r,--relative] <destination>
```

 see https://github.com/Quansight-Labs/jupyter-a11y-testing/actions/runs/4439607697/jobs/7792297908#step:4:4042

I tried multiple things, and the only thing that worked was adding that bit 🤷🏽‍♀️ 
see https://classic.yarnpkg.com/lang/en/docs/cli/link/

we can try and dig the bug now or merge and check later - but I am also not sure if this happened due to some changes in lumino itself

File: .github/workflows/accessibility-test-jupyterlab.yml (hunk 6 above)
Comment (trallard) [reply .github/workflows/accessibility-test-jupyterlab.yml]: Ok will revert then

File: README.md (hunk 7 above)
Comment (trallard) [reply README.md]: yes we need to discuss this on Monday I once again was not sure when creating the tree for the repo - will leave it as is now and we can make a decision on Monday

**Comment by trallard** [comment]
@gabalafou fixed the issues - except for the `jlpm link` bit

**Pull request merged**